### PR TITLE
Default to no-auth when credentials are absent

### LINKED ISSUE #62: Default to no-auth when credentials are not configured
## Background

Alternator deployments often run without authentication. The Rust driver currently supports this through `AlternatorConfig::builder().allow_no_auth()`, which delegates to the underlying AWS SDK DynamoDB config builder.

That works, but it makes the no-credentials case opt-in even though it is a common Alternator setup. Most examples that do not configure credentials need to remember to add `.allow_no_auth()` explicitly.

## Current behavior

A client configured with an Alternator endpoint but no credentials provider still needs an explicit `.allow_no_auth()` call. Without that, the underlying AWS SDK authentication path can require credentials even when the target Alternator server would accept unsigned requests.

Relevant code:

- `src/config.rs`: exposes `allow_no_auth()` / `set_allow_no_auth()`
- `src/client.rs`: builds the inner DynamoDB client and defaults `optimize_headers`, request compression, and region, but does not make no-auth the default

## Requested behavior

Make no-auth the default when credentials are not configured.

Clients that provide credentials should continue to sign requests exactly as they do today. This should only remove the need for explicit `.allow_no_auth()` in the no-credentials case.

## Acceptance criteria

- A client built with only an endpoint URL and no credentials provider can send Alternator requests without calling `.allow_no_auth()`.
- A client built with a credentials provider still sends signed requests.
- Existing explicit `.allow_no_auth()` remains valid and does not break compatibility.
- Tests cover both no-credentials and credentials-present behavior.
- README/examples can be simplified where they currently require `.allow_no_auth()` only because no credentials are configured.

### LINKED ISSUE #63: Do not preserve Authorization and X-Amz-Date headers when no credentials are configured
## Background

`optimize_headers` strips HTTP headers that Alternator does not need. The current Rust implementation uses a fixed allowlist in `src/optimize_headers.rs`:

- `host`
- `x-amz-target`
- `content-length`
- `accept-encoding`
- `content-encoding`
- `authorization`
- `x-amz-date`

The analogous Go driver builds its default allowlist conditionally: it keeps `Authorization` and `X-Amz-Date` only when credentials are configured.

## Current behavior

The Rust driver always preserves `authorization` and `x-amz-date` when `optimize_headers` is enabled, even for no-auth clients.

In the normal no-auth path these headers may simply be absent, so this often has no visible effect. But if they are present for any reason, `optimize_headers` allows them through even though there are no credentials and the request should be treated as unsigned.

## Requested behavior

When no credentials are configured / the request is using no-auth, header optimization should use a no-auth allowlist and drop `authorization` and `x-amz-date`.

When credentials are configured and the request is signed, header optimization should continue to preserve `authorization` and `x-amz-date`.

## Acceptance criteria

- With no credentials and header optimization enabled, outgoing requests do not preserve `authorization` or `x-amz-date`.
- With credentials and header optimization enabled, outgoing requests still preserve `authorization` and `x-amz-date`.
- Compression headers (`content-encoding`, `content-length`) remain preserved in both modes.
- Tests cover both no-auth and signed requests.
- Documentation reflects the conditional allowlist behavior.

Note: the requested header is `x-amz-date`; this issue uses that spelling even if callers sometimes typo it as `z-amz-date`.

### PR DESCRIPTION
**Pull request by dkropachev**
## Summary
- default Alternator clients without a credentials provider to SDK no-auth
- preserve signed per-request credentials after SDK signing, including plain SDK `config_override` credentials
- add explicit `require_auth()` for no-default-credentials clients that must fail instead of falling back to no-auth when per-request credentials are missing
- reject unsupported custom auth schemes/resolvers/preferences with an Alternator-specific message because Alternator supports only SigV4 static credentials or no-auth
- keep compression headers in both optimized allowlists and simplify no-auth README examples

Fixes #62
Fixes #63

## Testing
- `cargo fmt -- --check`
- `cargo test --lib`
- `docker compose up -d --wait`
- `cargo test --test http_content_tests optimize_headers -- --nocapture`
- `cargo test`
- `docker compose down --remove-orphans -v`

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -527,9 +532,13 @@ impl AlternatorBuilder {
 
 impl From<&aws_types::sdk_config::SdkConfig> for AlternatorBuilder {
     fn from(sdk_config: &aws_types::sdk_config::SdkConfig) -> Self {
+        let has_credentials_provider = sdk_config.credentials_provider().is_some();
```
Comment (m-szymon) [inline src/config.rs:603]: https://docs.rs/aws-sdk-dynamodb/latest/aws_sdk_dynamodb/struct.Config.html#method.credentials_provider:

> pub fn [credentials_provider](https://docs.rs/aws-sdk-dynamodb/latest/aws_sdk_dynamodb/struct.Config.html#method.credentials_provider)(&self) -> [Option](https://doc.rust-lang.org/nightly/core/option/enum.Option.html)<[SharedCredentialsProvider](https://docs.rs/aws-sdk-dynamodb/latest/aws_sdk_dynamodb/config/struct.SharedCredentialsProvider.html)>
👎Deprecated:
This function was intended to be removed, and has been broken since release-2023-11-15 as it always returns a None. Do not use.

I have seen your issues and started investigating. This line always is None and this shows that your changes and tests do not add anything new regarding header stripping.
We already had tests that in normal case of no-auth client Authorization is not added.
I am looking for a reason when this unnecessarily added and if we can detect it reliably.
I am not convinced that this is necessary.

Comment (dkropachev) [reply src/config.rs:603]: Thanks, good catch. I updated the PR to make this explicit:

- added unit tests around `aws_types::SdkConfig` conversion, so the credentials-provider detection is covered directly and does not rely on the deprecated `aws_sdk_dynamodb::Config::credentials_provider()` API;
- added an HTTP integration test that injects `authorization` and `x-amz-date` before the Alternator header-stripping interceptor runs, then verifies a no-auth optimized request drops them before the proxy sees the request.

This should cover both the credential-detection part and the case where auth headers are present for a no-auth client.

Comment (dkropachev) [reply src/config.rs:603]: I checked the env-credentials case separately on current `main` with an isolated environment: dummy `AWS_ACCESS_KEY_ID` / `AWS_SECRET_ACCESS_KEY`, empty temp AWS config/credentials files, temp `HOME`, and `AWS_EC2_METADATA_DISABLED=true`.

With `AlternatorConfig::builder()` and no explicit credentials provider, the request does not reach the wire at all. It fails before dispatch with `NoMatchingAuthScheme` / `NoIdentityResolver`, so no `Authorization` header is generated from env credentials.

So #62 is not about avoiding env credential signing. It is about making the common Alternator no-auth setup work by default: today, a client with no explicit credentials also needs `.allow_no_auth()`, otherwise it fails before sending.

For #63, I agree the normal no-auth path does not generate `Authorization`. The added injected-header integration test covers the edge case from the issue: if `Authorization` / `X-Amz-Date` are present for any reason before our header-stripping interceptor runs, the no-auth optimized path drops them, while signed clients still preserve them.

**Review by m-szymon — CHANGES_REQUESTED** [review]
In general we were quite focused on keeping compatibility with AWS driver, so that migration would be very easy - adding our features shouldn't have side effects on existing features. Even if user has his own customizations.
Auth in SDK is very generic and honestly I don't comprehend it fully. I am a little concerned if there are any realistic current or future user scenarios (e.g. some use of .push_auth_scheme(...).auth_scheme_resolver(...).auth_scheme_preference(...)) that would interfere here.

Two examples are connected to per-request overrides:

### Example 1: Signed Per-Request Credentials Can Be Stripped
A shared client may intentionally have no global credentials, while credentials are supplied per operation:
```rust
let client = AlternatorClient::from_conf(
    AlternatorConfig::builder()
        .endpoint_url(endpoint)
        .behavior_version_latest()
        .build(),
);

client
    .put_item()
    .customize()
    .config_override(
        aws_sdk_dynamodb::Config::builder()
            .credentials_provider(request_credentials)
            .region(aws_sdk_dynamodb::config::Region::new("us-east-1")),
    )
    .send()
    .await?;
```
The SDK can select SigV4 and add Authorization / X-Amz-Date, but AlternatorInterceptor decides whether to preserve auth headers from client construction state. Since the client had no credentials provider then, optimized header stripping can remove the SigV4 headers and break a request that works with pure aws_sdk_dynamodb.

### Example 2: Missing Per-Request Credentials Become Unsigned Requests
For the same request-scoped credential pattern, a missing per-operation override used to fail auth resolution locally with the AWS SDK:
```rust
client
    .put_item()
    // forgot `.customize().config_override(...credentials_provider...)`
    .send()
    .await?;
```
After automatic allow_no_auth(), the request can fall back to noAuth and be sent unsigned. That changes a missing-credentials bug from an SDK error into a successful unsigned request against no-auth Alternator deployments.

**Comment by dkropachev** [comment]
> In general we were quite focused on keeping compatibility with AWS driver, so that migration would be very easy - adding our features shouldn't have side effects on existing features. Even if user has his own customizations. Auth in SDK is very generic and honestly I don't comprehend it fully. I am a little concerned if there are any realistic current or future user scenarios (e.g. some use of .push_auth_scheme(...).auth_scheme_resolver(...).auth_scheme_preference(...)) that would interfere here.

In other drivers we raise (whenever we can) when customization is not compatible with alternator.
Currently alternator supports only static credentials, it does not support STA or any kind of temporary token and most likely will never support them.
Also by default alternator authentication is disabled and which is also make sense because Auth header takes more than 500 bytes from request.
That is why in other drivers we make sdk not to pick up credentials from the environment and rather supply them in the client config and by default it sends no credentials.
This has two purposes:
1. We can't don't send extra header and show better perfromance by default
2. If client wants credentials they supply it explicitly, so we get to see it in the code
3. Alternator and dynamodb infra is not mixed up, so you can have alternator and dynamodb clients working together in the same binary, which is great for migration.

> 
> Two examples are connected to per-request overrides:
> 
> ### Example 1: Signed Per-Request Credentials Can Be Stripped
> A shared client may intentionally have no global credentials, while credentials are supplied per operation:
> 
> ```rust
> let client = AlternatorClient::from_conf(
>     AlternatorConfig::builder()
>         .endpoint_url(endpoint)
>         .behavior_version_latest()
>         .build(),
> );
> 
> client
>     .put_item()
>     .customize()
>     .config_override(
>         aws_sdk_dynamodb::Config::builder()
>             .credentials_provider(request_credentials)
>             .region(aws_sdk_dynamodb::config::Region::new("us-east-1")),
>     )
>     .send()
>     .await?;
> ```

True customization API spoils the whole thing, we can address some of it by failing prematurely and allow only supported customizations.

> 
> The SDK can select SigV4 and add Authorization / X-Amz-Date, but AlternatorInterceptor decides whether to preserve auth headers from client construction state. Since the client had no credentials provider then, optimized header stripping can remove the SigV4 headers and break a request that works with pure aws_sdk_dynamodb.
> 
> ### Example 2: Missing Per-Request Credentials Become Unsigned Requests
> For the same request-scoped credential pattern, a missing per-operation override used to fail auth resolution locally with the AWS SDK:
> 
> ```rust
> client
>     .put_item()
>     // forgot `.customize().config_override(...credentials_provider...)`
>     .send()
>     .await?;
> ```
> 
> After automatic allow_no_auth(), the request can fall back to noAuth and be sent unsigned. That changes a missing-credentials bug from an SDK error into a successful unsigned request against no-auth Alternator deployments.

This is fine, it should be guarded on alternator side, if they allow anonymouse operation it is on em and if they forgot credential it is also on em, we can't and should not guard it.

**Comment by dkropachev** [comment]
@m-szymon , I also pushed another commit with dropping `auth_scheme_preference` to avoid confusion, if we will get it needed there will be no problem to add it back.

File: tests/http_content/optimize_headers.rs [hunk 2]
```diff
@@ -50,6 +66,30 @@ use aws_sdk_dynamodb::types::{
 
 use alternator_driver::*;
 
+fn request_credentials() -> aws_sdk_dynamodb::config::Credentials {
```
Comment (m-szymon) [inline tests/http_content/optimize_headers.rs:69]: nit: tests, even in this file already use `aws_sdk_dynamodb::config::Credentials::for_tests_with_session_token()`.
Which BTW is misleading as we don't support temporary credentials in Alternator, so `aws_sdk_dynamodb::config::Credentials::for_tests()` would be better.
Let's be consistent.

**Review by m-szymon — APPROVED** [review]
Ok.
However whole README may require some review. Claims about “seamless swap” and “identical interface” start sounding to strong. Maybe they require some note that it applies to features supported by Alternator and we don't guarantee that it still works with DynamoDB.
Or maybe we should have a "migration" client version that is strictly compatible with DynamoDB?

**Comment by dkropachev** [comment]
Addressed in 6a0458e: softened the README compatibility wording and switched the optimize_headers tests to static test credentials.

File: tests/http_content/optimize_headers.rs (hunk 2 above)
Comment (dkropachev) [reply tests/http_content/optimize_headers.rs:69]: Done in 6a0458e. Switched the optimize_headers fixtures to Credentials::for_tests() for consistency with Alternator static-credentials support.
